### ISSUE TITLE
Detection Accuracy 

### ISSUE THREAD

**Issue by alic-xc** [OlafenwaMoses/ImageAI#665]
Hi guys,
i have this project i'm working on, what it does is that it detect and count the total fish in a picture. After training, i tried to get the total fish in this picture but it is not giving me the accurate value. Meanwhile, this picture only contain small number of fish, i have pictures that contain higher number
![earlyFingerling83](https://user-images.githubusercontent.com/23708960/118776685-2752e200-b880-11eb-8bc2-aed1106f47cb.jpg)
![earlyFingerling152](https://user-images.githubusercontent.com/23708960/118777520-1060bf80-b881-11eb-9b13-5efe272df4f2.jpg)

Any help will be appreciated!

**Comment by ekesdf** [comment]
Hi can you Show Dome Images with drawn bboxes thy so I can get a rough idea how "accurate" your model is

**Comment by alic-xc** [comment]
> Hi can you Show Dome Images with drawn bboxes thy so I can get a rough idea how "accurate" your model is

![earlyFingerling700-detected](https://user-images.githubusercontent.com/23708960/118801459-f29f5480-b898-11eb-84d9-70dc9f0aa3f9.jpg)

This is an example

**Comment by ekesdf** [comment]
I see what your problem is all the labels you have mad are rectangle and probably have other part from other fish like 
the one I made in this picture 
![118777520-1060bf80-b881-11eb-9b13-5efe272df4f2](https://user-images.githubusercontent.com/64316659/118807846-166a9680-b8a9-11eb-83bc-61c07a799b27.jpg)
so this way of annotating actually trains your model to detect many fishes at the same time bc your dataset includes bbox with more than one fish like a fish head or  fish body.  And some get detected correctly bc you also have some bbox with a clear white space around it. So for that particular project I would give you the advice to train a segmentation model. It accommodates for those weird shape, and it will probably work better for your case it prevents from "overlapping" bboxes

**Comment by alic-xc** [comment]
To understand what you mean, the best way to achieve good accuracy is through the use of image segmentation. Thanks so much.
Can you provide a brief guide which i can follow to this image segmentation, train model or can ImageAI also do that ?

**Comment by OlafenwaMoses** [comment]
@alic-xc if you are considering using Image Segmentation ( best to use instance segmentation ), you can use this library.

https://github.com/ayoolaolafenwa/PixelLib

**Comment by ekesdf** [comment]
@alic-xc  did it work to train the segmentation model ?
and how are the results ?

**Comment by alic-xc** [comment]
@ekesdf the result is very poor. I just had to stop the project because it wasted my time and waste resources. I don't know whether it was the library I used or the model I generated. 
However I'm still open to any better approach one can use to solve the problem.

**Comment by ekesdf** [comment]
Do you have your dataset uploaded somewhere, so I can try to get it running ?

**Comment by alic-xc** [comment]
@ekesdf Okay.
I used pixelLib (https://github.com/ayoolaolafenwa/PixelLib) for it.
this is the link for the trained models (https://drive.google.com/drive/folders/1OKhxbJ-tCddcm3acq7beeobOvn6DJR6x?usp=sharing)
for the dataset (https://drive.google.com/file/d/1Q8OYH1BCjTpd4vYvumy4wOZ9j6fLVApk/view?usp=sharing)

**Comment by ekesdf** [comment]
@alic-xc good new for you, I was able to train a model that can predict correctly like shown in this picture

**Prediction**
![test](https://user-images.githubusercontent.com/64316659/122684481-e4c95000-d205-11eb-87e3-774aa19c9715.jpg)
**Original**
![earlyFingerling260](https://user-images.githubusercontent.com/64316659/122684502-fca0d400-d205-11eb-81e1-932eedf7c343.jpg)

Here is the [model](https://drive.google.com/file/d/1Vjo1ielAb4Gb7YkgI5j908EjiKP0YqmE/view?usp=sharing) with the best acc

tensorflow==2.4.0
cuda==11.0
cudnn==8.0.5.39-1
python==3.8.5-64bit


**Code used for training**
```
from pixellib.custom_train import instance_custom_training

train_maskrcnn = instance_custom_training()
train_maskrcnn.modelConfig(network_backbone = "resnet101", num_classes= 1, batch_size = 2,gpu_count=1)
train_maskrcnn.load_pretrained_model("mask_rcnn_coco.h5")
train_maskrcnn.load_dataset("fingerlings")

train_maskrcnn.train_model(num_epochs = 10, augmentation=True,  path_trained_models = "mask_rcnn_coco")
```
**Code used for prediction**
```
from pixellib.instance import custom_segmentation

test_video = custom_segmentation()
test_video.inferConfig(num_classes=  1, class_names=["fish"])
test_video.load_model("/mask_rcnn_model.066-0.176888.h5")
test_video.segmentImage("fingerlings/test/earlyFingerling368.jpg",output_image_name="test.jpg")
```
**Code used for evaluation**
```
from pixellib.custom_train import instance_custom_training

train_maskrcnn = instance_custom_training()
train_maskrcnn.modelConfig(network_backbone = "resnet101", num_classes= 1)
train_maskrcnn.load_dataset("fingerlings")
train_maskrcnn.evaluate_model("mask_rcnn_model.010-0.184201.h5",iou_threshold=0.8)

```

**Comment by alic-xc** [comment]
@ekesdf Wow, that's a great work you did. 
However, i haven't test it with multiple fishes in an image. Let me test and get back to you.

**Comment by ekesdf** [comment]
With the code I send it the previous comment the training should work fine on any data 

Like with one fish or 2 or more in the same picture

**Comment by alic-xc** [comment]
Good. i'm testing it now, will give you feedback on that.

**Comment by alic-xc** [comment]
@ekesdf I tested it yesterday, and it was working fine. I have a ton of question but I would like to send it to you personally, so I can close this issue.
Would you mind sharing the easy way to reach you?

**Comment by ekesdf** [comment]
I am gonna Open an issue at one of your repos 
With the information

**Comment by alic-xc** [comment]
That's fine.
